### PR TITLE
Simplify the Text API.

### PR DESCRIPTION
**Pull request by jpountz**
We have the Text API, which is essentially a wrapper around a String and a
BytesReference and then we have 3 implementations depending on whether the
String view should be cached, the BytesReference view should be cached, or both
should be cached.

This commit merges everything into a single Text that is essentially the old
StringAndBytesText impl.

Long term we should look into whether this API has any performance benefit or
if we could just use plain strings. This would greatly simplify all our other
APIs that currently use Text.

### CODE REVIEW

File: modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/SimpleSortTests.java [hunk 1]
```diff
@@ -1650,8 +1650,8 @@ public void testNestedSort() throws IOException, InterruptedException, Execution
             assertThat(hits[i].getSortValues().length, is(1));
             Object o = hits[i].getSortValues()[0];
             assertThat(o, notNullValue());
-            assertThat(o instanceof StringAndBytesText, is(true));
-            StringAndBytesText text = (StringAndBytesText) o;
+            assertThat(o instanceof Text, is(true));
```
Comment (nik9000) [inline modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/SimpleSortTests.java]: I know you've just copied but is it worth even asserting this? Also, if you do assert this maybe use `assertThat(o, instanceOf(Text.class))`;

**Comment by nik9000** [comment]
LGTM. Left small comment. Having two versions might have been useful if they were immutable but it looks like none were. So this doesn't change the immutable-ness anywhere so its safe and simpler.

File: modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/SimpleSortTests.java (hunk 1 above)
Comment (jpountz) [inline modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/SimpleSortTests.java]: I think it's worth doing given that getSortValues() returns an Object[]?

Comment (nik9000) [inline modules/lang-groovy/src/test/java/org/elasticsearch/messy/tests/SimpleSortTests.java]: I'm just used to `ClassCastException`s  in tests being test failures. The cast is implicitly part of the test. I'm fine with keeping the explicit assertion but I'd prefer the other style so you get nicer error messages.

**Comment by jpountz** [comment]
Thanks @nik9000 for having a look. I just pushed a new commit.

**Comment by nik9000** [comment]
LGTM

**Pull request merged**